File: Src/Compilers/VisualBasic/Portable/Compilation/LookupSymbolsInfo.vb
' Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
    Friend NotInheritable Class LookupSymbolsInfo
        Inherits AbstractLookupSymbolsInfo(Of Symbol)

        ' TODO: tune pool size
        Private Shared ReadOnly poolSize As Integer = 64
        Private Shared ReadOnly pool As New ObjectPool(Of LookupSymbolsInfo)(Function() New LookupSymbolsInfo(), poolSize)

        Private Sub New()
            MyBase.New(IdentifierComparison.Comparer)
        End Sub

        ' To implement Poolable, you need two things:
        ' 1) Expose Freeing primitive. 
        Public Sub Free()
            ' Note that poolables are not finalizable.  If one gets collected - no big deal.
            Me.Clear()
            pool.Free(Me)
        End Sub

        Public Shared Function GetInstance() As LookupSymbolsInfo
            Dim info As LookupSymbolsInfo = pool.Allocate()
            Debug.Assert(info.Names.Count = 0)
            Return info
        End Function
    End Class
End Namespace

File: src/Compilers/CSharp/Portable/Binder/LookupSymbolsInfo.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#nullable disable

using System;
using System.Diagnostics;
using Microsoft.CodeAnalysis.PooledObjects;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.CSharp
{
    internal sealed class LookupSymbolsInfo : AbstractLookupSymbolsInfo<Symbol>
    {
        // TODO: tune pool size.
        private const int poolSize = 64;
        private static readonly ObjectPool<LookupSymbolsInfo> s_pool = new ObjectPool<LookupSymbolsInfo>(() => new LookupSymbolsInfo(), poolSize);

        private LookupSymbolsInfo()
            : base(StringComparer.Ordinal)
        {
        }

        // To implement Poolable, you need two things:
        // 1) Expose Freeing primitive. 
        public void Free()
        {
            // Note that poolables are not finalizable. If one gets collected - no big deal.
            this.Clear();
            s_pool.Free(this);
        }

        // 2) Expose the way to get an instance.
        public static LookupSymbolsInfo GetInstance()
        {
            var info = s_pool.Allocate();
            Debug.Assert(info.Count == 0);
            return info;
        }
    }
}
